### PR TITLE
Move allowed_actions and default_action up to Resource

### PR DESCRIPTION
**Pull request by jkeiser**

### CODE REVIEW

File: lib/chef/resource.rb [hunk 1]
```diff
@@ -943,8 +954,63 @@ def provider_base(arg=nil)
         end
         @provider_base ||= arg || Chef::Provider
       end
-    end
 
+      #
```
Comment (danielsdeleo) [inline lib/chef/resource.rb]: Would you mind changing this from `class << self` to `self.method` style? Took me too much work to figure out why there are two allowed actions methods. There is also some weird stuff with ivar setting and method lookup that happens in `class << self`, which isn't a problem yet, but recently bit me on a different project.

File: lib/chef/resource.rb [hunk 2]
```diff
@@ -943,8 +954,63 @@ def provider_base(arg=nil)
...
+      # The list of allowed actions for the resource.
+      #
+      # @param actions [Array<Symbol>] The list of actions to add to allowed_actions.
+      #
+      # @return [Arrau<Symbol>] The list of actions, as symbols.
```
Comment (danielsdeleo) [inline lib/chef/resource.rb]: typo

File: lib/chef/resource.rb (hunk 1 above)
Comment (jkeiser) [inline lib/chef/resource.rb]: Sure!

File: DOC_CHANGES.md [hunk 3]
```diff
@@ -6,19 +6,24 @@ Example Doc Change:
 Description of the required change.
 -->
 
-### Resources must now use `provides` to declare recipe DSL
+### Resources must now use `resource_name` (or `provides`) to declare recipe DSL
 
 Resources declared in `Chef::Resource` namespace will no longer get recipe DSL
-automatically.  Instead, explicit `provides` is required in order to have DSL:
+automatically.  Instead, `resource_name` is required in order to have DSL:
```
Comment (randomcamel) [inline DOC_CHANGES.md]: Line above says `resource_name` or `provides`, this line only says `resource_name`.

File: lib/chef/resource/bff_package.rb [hunk 4]
```diff
@@ -22,13 +22,7 @@
...
   class Resource
     class BffPackage < Chef::Resource::Package
-      provides :bff_package
-
-      def initialize(name, run_context=nil)
-        super
-        @resource_name = :bff_package
-      end
-
+      use_automatic_resource_name
```
Comment (randomcamel) [inline lib/chef/resource/bff_package.rb]: Mmm less boilerplate.

File: spec/unit/resource_spec.rb [hunk 5]
```diff
@@ -336,6 +332,85 @@ class ResourceTestHarness < Chef::Resource
     end
   end
 
+  describe "self.resource_name" do
+    context "When resource_name is not set" do
+      it "and there are no provides lines, resource_name is nil" do
+        c = Class.new(Chef::Resource) do
+        end
+
+        r = c.new('hi')
```
Comment (randomcamel) [inline spec/unit/resource_spec.rb]: I can't help but think `c` and `r` are unnecessarily brief variable names.

**Comment by randomcamel** [comment]
:+1: overall, though I'd like to look it over a final time after a rebase.

**Comment by jkeiser** [comment]
This PR was against jk/resource_name.  Recreating against master.

**Pull request closed without merging**